### ISSUE TITLE
Does not work if match value comes from a variable

### ISSUE THREAD

**Issue by francisbrito** [z-pattern-matching/z#37]
e.g:

```js
#!/usr/bin/env node
const { matches } = require('z')

const FILTERS = {
  CATEGORY: 'CATEGORY',
  PRICE: 'PRICE',
  COLOR: 'COLOR',
  ONLY_NEW: 'ONLY_NEW',
  ONLY_IN_DISCOUNT: 'ONLY_IN_DISCOUNT'
}

const _30DaysAgo = Date.now() - 1000 * 60 * 60 * 24 * 30
const convertOptionsToQuery = filterOptions => matches(filterOptions.by)(
  (b = FILTERS.CATEGORY) => ({ category: filterOptions.value }),
  (b = FILTERS.COLOR) => ({ colors: filterOptions.value }),
  (b = FILTERS.PRICE) => ({ price: { $gte: filterOptions.value[0], $lte: filterOptions.value[1] } }),
  (b = FILTERS.ONLY_NEW) => ({ createdAt: { $gte: new Date(_30DaysAgo) } }),
  (b = FILTERS.ONLY_IN_DISCOUNT) => ({ isInDiscount: true }),
  (b) => ({})
)

convertOptionsToQuery({ by: 'CATEGORY', value: 'shirts' })

// ReferenceError: FILTERS is not defined
```

Strangely enough, it only happens if the code is ran as a file, if it is ran in the REPL it works.

This error seems to be related to `js-function-reflector`.

**Comment by leonardiwagner** [comment]
@francisbrito This code won't work because `FILTERS` is not present in z `matches` scope. That's why we can only make a pattern with explicit values. It's actually strange how it works on REPL, I'm going to spend some time to check this.

If someone has a idea how could inside `matches` read variables outside its scope, it would be a great feature! Honestly I don't know any solution but having a featue to pass the `FILTERS` variable explicitly to `z`.

**Comment by leonardiwagner** [comment]
Hello @francisbrito, this feature is now available on z@1.0.4 , now you can call z with `apply` or `call` and then set the context, examples:

```javascript
    const FILTERS = {
      CATEGORY: 'CATEGORY',
      PRICE: 'PRICE',
      COLOR: 'COLOR',
      ONLY_NEW: 'ONLY_NEW',
      ONLY_IN_DISCOUNT: 'ONLY_IN_DISCOUNT'
    }
    
    const result = matches(FILTERS.COLOR).call(this,
      (x = FILTERS.CATEGORY) => false,
      (x = FILTERS.COLOR) => true,
      (x = FILTERS.ONLY_NEW) => false
    )
```

If the variable to match isn't on `this`, you can also set manually , eg.:

```javascript
    const FILTERS = {
      CATEGORY: 'CATEGORY',
      PRICE: 'PRICE',
    }
    
    const SORT = {
      ASC: 'ASC',
      DESC: 'DESC',
    }

    const result = matches('something').call({ FILTERS, SORT },
      (x = FILTERS.CATEGORY) => false,
      (x = SORT.ASC) => false
    )
```

This was possible due [a modification on js-function-reflector library](https://github.com/arrizalamin/js-function-reflector/pull/12)

**Comment by francisbrito** [comment]
Glad hearing so. Thank you very much!
